### PR TITLE
Use constant

### PR DESCRIPTION
**Pull request by Cjkjvfnby**
It is based on previous PR

I replace True - False to string constants. This make cache more readable.

### CODE REVIEW

File: default/AI/ShipDesignAI.py [hunk 1]
```diff
@@ -77,8 +77,12 @@
 MISSING_REQUIREMENT_MULTIPLIER = -1000
 INVALID_DESIGN_RATING = -999  # this needs to be negative but greater than MISSING_REQUIREMENT_MULTIPLIER
 
+
+WITH_UPKEEP = 'with upkeep'
+WITHOUT_UPKEEP = 'without outkeep'
```
Comment (Morlic-fo) [inline default/AI/ShipDesignAI.py:82]: Spelling. Also "(not) considering fleet upkeep" is more true to the meaning I think.

**Comment by Cjkjvfnby** [comment]
I try to simplify print_best_designs:
- use dict.items() 
- print with prefix, not suffix

I am stop for today.  Here is my result. 

``` python
    def print_best_designs(self):
        """Print the best designs that were previously found."""
        print "Currently cached best designs:"
        for classname in self.best_designs:
            print classname
            cache_name = self.best_designs[classname]
            for consider_fleet, cache_upkeep in cache_name.items():
                print 4*" ", '%s:' % consider_fleet
                for req_tuple, cache_reqs in cache_upkeep.items():
                    print 8*" ", 'requirements: %s:' % ', '.join(req_tuple)
                    for tech_tuple, cache_techs in cache_reqs.items():
                        print 12*" ", "relevant tech updates: %s:" % ', '.join(tech_tuple) or None
                        for species_tuple, cache_species in cache_techs.items():
                            print 16*" ", 'relevant species stats: %s:' % ', '.join(species_tuple) or None
                            for av_parts, cache_parts in cache_species.items():
                                print 20*" ", 'av_parts %s:' % ', '.join(sorted(av_parts))
                                for hullname in sorted(cache_parts, reverse=True, key=lambda x: cache_parts[x][0]):
                                    print 24*" ", '%s: %.4f parts: %s' % (hullname,
                                                                          cache_parts[hullname][0],
                                                                          ', '.join(cache_parts[hullname][1])
                                                                          )
```

**Comment by Morlic-fo** [comment]
Not sure if it is simpler but it is shorter using dict.items(). I guess first for loop can be rewritten the same way.

> print with prefix, not suffix

Not sure what you mean with this. Do you mean switching for example
`print 12*" ", tech_tuple, " # relevant tech upgrades"`
to
``print 12*" ", "relevant tech updates: %s:" % ', '.join(tech_tuple) or None`

I would actually argue that the cache is much easier to read in logfile if the line starts with the cache-content and then is explained by a `#comment`. That way when reading you do not have to skim to find the point in the line where the relevant data is  and what is "useless" explaining code once you are used to the structure:
`/*Some lost comment before the code*/ makes_code.hard_to_read()`
`this_code.is_easy_to_read() // Because comment is where it should be`

I am open for rephrasing/reformatting the comment/explanation string but please keep it at the end of the line.

**Comment by Cjkjvfnby** [comment]
Here is output example and code.
Please check bottom of output for  `SH_ASTEROID: -999.0000; parts: , , , , ,` it looks strange 

```
MilitaryShipDesigner
     with upkeep:
         minFuel: 1, minSpeed: 30, enemyDmg: 0, enemyShields: 0.0, enemyMineDmg: 0:  # requirements
             None:  # relevant tech updates
                 WEAPON: BAD, SHIELDS: :  # relevant species stats
                     AR_STD_PLATE, FU_BASIC_TANK, SR_WEAPON_1_1:  # available parts
                         SH_STANDARD: 1.1766; parts: SR_WEAPON_1_1, SR_WEAPON_1_1, AR_STD_PLATE, 
                         SH_BASIC_MEDIUM: 0.7389; parts: SR_WEAPON_1_1, AR_STD_PLATE, 
                         SH_BASIC_SMALL: 0.3656; parts: SR_WEAPON_1_1
                         SH_COLONY_BASE: -30000.0000; parts: FU_BASIC_TANK
         minFuel: 1, minSpeed: 30, enemyDmg: 5.0, enemyShields: 0.0, enemyMineDmg: 0:  # requirements
             None:  # relevant tech updates
                 WEAPON: BAD, SHIELDS: :  # relevant species stats
                     AR_STD_PLATE, FU_BASIC_TANK, SR_WEAPON_1_1:  # available parts
                         SH_STANDARD: 1.2504; parts: SR_WEAPON_1_1, SR_WEAPON_1_1, AR_STD_PLATE, 
                         SH_BASIC_MEDIUM: 0.7779; parts: SR_WEAPON_1_1, AR_STD_PLATE, 
                         SH_BASIC_SMALL: 0.3822; parts: SR_WEAPON_1_1
                         SH_COLONY_BASE: -30000.0000; parts: FU_BASIC_TANK
                     FU_BASIC_TANK:  # available parts
                         SH_COLONY_BASE: -30000.0000; parts: FU_BASIC_TANK
             SHP_WEAPON_1_2:  # relevant tech updates
                 WEAPON: BAD, SHIELDS: :  # relevant species stats
                     AR_STD_PLATE, FU_BASIC_TANK, SR_WEAPON_1_1:  # available parts
                         SH_STANDARD: 1.8755; parts: SR_WEAPON_1_1, SR_WEAPON_1_1, AR_STD_PLATE, 
                         SH_BASIC_MEDIUM: 1.1668; parts: SR_WEAPON_1_1, AR_STD_PLATE, 
                         SH_BASIC_SMALL: 0.5734; parts: SR_WEAPON_1_1
                         SH_COLONY_BASE: -30000.0000; parts: FU_BASIC_TANK
                     FU_BASIC_TANK:  # available parts
                         SH_COLONY_BASE: -30000.0000; parts: FU_BASIC_TANK
     without outkeep:
         minFuel: 1, minSpeed: 30, enemyDmg: 0, enemyShields: 0.0, enemyMineDmg: 0:  # requirements
             None:  # relevant tech updates
                 WEAPON: BAD, SHIELDS: :  # relevant species stats
                     AR_STD_PLATE, FU_BASIC_TANK, FU_TRANSPATIAL_DRIVE, SR_WEAPON_1_1:  # available parts
                         SH_STANDARD: 1.0354; parts: SR_WEAPON_1_1, SR_WEAPON_1_1, AR_STD_PLATE, 
                         SH_BASIC_MEDIUM: 0.6634; parts: SR_WEAPON_1_1, AR_STD_PLATE, 
                         SH_TRANSSPATIAL: 0.6098; parts: SR_WEAPON_1_1, 
                         SH_BASIC_SMALL: 0.3331; parts: SR_WEAPON_1_1
                         SH_COLONY_BASE: -30000.0000; parts: FU_BASIC_TANK
                     AR_STD_PLATE, FU_BASIC_TANK, SR_WEAPON_1_1:  # available parts
                         SH_ASTEROID: 2.3179; parts: SR_WEAPON_1_1, SR_WEAPON_1_1, SR_WEAPON_1_1, AR_STD_PLATE, , FU_BASIC_TANK
                         SH_SMALL_ASTEROID: 1.0399; parts: SR_WEAPON_1_1, 
                         SH_STANDARD: 1.0354; parts: SR_WEAPON_1_1, SR_WEAPON_1_1, AR_STD_PLATE, 
                         SH_BASIC_MEDIUM: 0.6634; parts: SR_WEAPON_1_1, AR_STD_PLATE, 
                         SH_BASIC_SMALL: 0.3331; parts: SR_WEAPON_1_1
                         SH_COLONY_BASE: -30000.0000; parts: FU_BASIC_TANK
         minFuel: 1, minSpeed: 30, enemyDmg: 5.0, enemyShields: 0.0, enemyMineDmg: 0:  # requirements
             None:  # relevant tech updates
                 WEAPON: BAD, SHIELDS: :  # relevant species stats
                     AR_STD_PLATE, FU_BASIC_TANK, FU_TRANSPATIAL_DRIVE, SR_WEAPON_1_1:  # available parts
                         SH_STANDARD: 1.0354; parts: SR_WEAPON_1_1, SR_WEAPON_1_1, AR_STD_PLATE, 
                         SH_BASIC_MEDIUM: 0.6634; parts: SR_WEAPON_1_1, AR_STD_PLATE, 
                         SH_TRANSSPATIAL: 0.6098; parts: SR_WEAPON_1_1, 
                         SH_BASIC_SMALL: 0.3331; parts: SR_WEAPON_1_1
                         SH_COLONY_BASE: -30000.0000; parts: FU_BASIC_TANK
                     FU_BASIC_TANK, FU_TRANSPATIAL_DRIVE:  # available parts
                         SH_TRANSSPATIAL: -999.0000; parts: , 
                         SH_COLONY_BASE: -30000.0000; parts: FU_BASIC_TANK
                     AR_STD_PLATE, FU_BASIC_TANK, SR_WEAPON_1_1:  # available parts
                         SH_ASTEROID: 2.3179; parts: SR_WEAPON_1_1, SR_WEAPON_1_1, SR_WEAPON_1_1, AR_STD_PLATE, , FU_BASIC_TANK
                         SH_SMALL_ASTEROID: 1.0399; parts: SR_WEAPON_1_1, 
                         SH_STANDARD: 1.0354; parts: SR_WEAPON_1_1, SR_WEAPON_1_1, AR_STD_PLATE, 
                         SH_BASIC_MEDIUM: 0.6634; parts: SR_WEAPON_1_1, AR_STD_PLATE, 
                         SH_BASIC_SMALL: 0.3331; parts: SR_WEAPON_1_1
                         SH_COLONY_BASE: -30000.0000; parts: FU_BASIC_TANK
                     FU_BASIC_TANK:  # available parts
                         SH_ASTEROID: -999.0000; parts: , , , , , 
                         SH_SMALL_ASTEROID: -999.0000; parts: , 
                         SH_COLONY_BASE: -30000.0000; parts: FU_BASIC_TANK
```

``` python
def print_best_designs(self):
        """Print the best designs that were previously found."""
        print "Currently cached best designs:"
        for classname, cache_name in self.best_designs.items():
            print classname
            for consider_fleet, cache_upkeep in cache_name.items():
                print 4*" ", '%s:' % consider_fleet
                for req_tuple, cache_reqs in cache_upkeep.items():
                    print 8*" ", '%s:  # requirements' % (', '.join(req_tuple) or None)
                    for tech_tuple, cache_techs in cache_reqs.items():
                        print 12*" ", "%s:  # relevant tech updates" % (', '.join(tech_tuple) or None)
                        for species_tuple, cache_species in cache_techs.items():
                            print 16*" ", '%s:  # relevant species stats' % (', '.join(species_tuple) or None)
                            for av_parts, cache_parts in cache_species.items():
                                print 20*" ", '%s:  # available parts' % ', '.join(sorted(av_parts))
                                for hullname in sorted(cache_parts, reverse=True, key=lambda x: cache_parts[x][0]):
                                    print 24*" ", '%s: %.4f; parts: %s' % (hullname,
                                                                           cache_parts[hullname][0],
                                                                           ', '.join(cache_parts[hullname][1])
                                                                           )
```

**Comment by Morlic-fo** [comment]
> Please check bottom of output for SH_ASTEROID: -999.0000; parts: , , , , , it looks strange

What is it in particular that you find strange? Just to make sure I don't miss something you mean.

My concern:
I am actually unsure how that branch is available. It seems to ignore any external slots of the Asteroid Hull. It may be a bug with the ResearchAI or some issue related to the queried locations. I will check it tomorrow.

**Pull request merged**